### PR TITLE
Feature/add ece calibration

### PR DESCRIPTION
**Pull request by DHdroid**
- Added calibration

Workflow 다시 정리
1) huggingface dataset 받는다 (jsonl, 안나오면 parquet 파일을 변환해야 함.)
`data_utils/generate_raw_dataset.ipynb` 사용

2) 1000 character 기준  row filter (특히, yelp)

3)  train / valid(dev) / test → 4번으로, valid가 없으면 → 4번으로

4)  정윤님이 올려주신 노트북으로 train에서 10%를 valid로 뺀다. (random seed 42, 앞으로 모든 random seed 42) `data_utils/generate_raw_dataset.ipynb` 사용
5)  train / valid / test → huggingface에 올린다. ICKD/sst5-raw

6)  raw dataset을 이용해서 bert 학습한다. 
이때 datamodule에 하드코딩되어 있는 huggingface dataset를 ICKD에 있는 것으로 교체한다. (ex. ag_news -> ICKD/agnews-raw)
batch size 64 고정, lr은 [1e-5, 2e-5, 5e-5]에서 탐색

7) 학습된 bert로 train / valid / test set inference 결과 만든다. → valid/test 포함해서 ICKD/sst5-bert

8) ECE를 minimize하는 temperature를 구해서, train / valid / test set inference 결과를 scale한다. → valid/test 포함해서 ICKD/sst5-bert-scaled
예시 명령어) `make temperature-scale DATASET=ICKD/agnews-bert SPLIT=valid OUTPUT_FOLDER=data/agnews/scaled`

### CODE REVIEW

File: scripts/infer_bert.py [hunk 1]
```diff
@@ -21,6 +21,7 @@
     infer(
         checkpoint_path=args.checkpoint_path,
         dataset=args.dataset,
+        dataset_split=args.dataset_split,
```
Comment (yoonichoi) [inline scripts/infer_bert.py:24]: 이 부분 혹시 infer 돌리실때 에러가 직접 떠서 수정하신건가요? 저는 이 부분 수정없이도 잘 돌아갔었거든요 (make infer-bert가 콜하는 scripts/infer_bert.py가 iclx/soft_label_generator/infer_bert.py의 infer function에 기반하고 있어서 그런지 dataset_split argument넣어줬을때 에러없이 잘 돌아갔었어서 확인부탁드릴게요

File: iclx/soft_label_generator/datamodule/ag_news.py
Comment (yoonichoi) [inline iclx/soft_label_generator/datamodule/ag_news.py:1]: calibration feature사용을 위해 다른 datamodule들도 이런식으로 바꿔야할까요? 아니면 그냥 동현님 ag_news작업하실때 바꾸신건가요? 후자라면 이 PR에는 일단 ece calibration 관련 수정사항만 넣는 것이 어떨까싶은 생각입니다! (datamodule load_dataset방식 수정때 한 번에 바꾸게끔)

File: scripts/temperature_scale.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+import argparse
+
+from iclx.soft_label_generator.calculate_temperature import calculate_temperature
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--infer_dataset_name", type=str) # dataset name from HF. ex) ICKD/agnews-bert-scaled
+    parser.add_argument("--target_split", type=str) # train, valid, test
```
Comment (yoonichoi) [inline scripts/temperature_scale.py:13]: just for clarification: 저희는 validation set로 temperature 찾는거죠? if so, why don't we just set it to be "valid" by default?

Comment (DHdroid) [reply scripts/temperature_scale.py:13]: [관련 논의](https://neurest.slack.com/archives/C0725QGAGH0/p1715247673554439?thread_ts=1715242050.885199&cid=C0725QGAGH0)에 의해 추가한 argument입니다!

File: scripts/infer_bert.py (hunk 1 above)
Comment (DHdroid) [reply scripts/infer_bert.py:24]: 이걸 안바꿔주면 valid, test split에 인퍼런스가 불가합니다.. [관련논의](https://neurest.slack.com/archives/C0725QGAGH0/p1715257461423169?thread_ts=1715257218.308449&cid=C0725QGAGH0)

File: iclx/soft_label_generator/datamodule/ag_news.py
Comment (DHdroid) [reply iclx/soft_label_generator/datamodule/ag_news.py:1]: 위 workflow 6번에 해당하는 내용이라 다른 분들도 따라갈 수 있게 하려고 수정사항에 포함시켰습니다.

File: pyproject.toml
Comment (nemodleo) [inline pyproject.toml:1]: a100 faiss-gpu 재설치함에 있어 poetry toml의 faiss-gpu을 제거하게 되는데요
default faiss-gpu(이전 poetry파일에 해당하는 버전 1.7.2)가 잘 작동하는 경우도 있기에
해당 업데이트는 빼야 할듯합니다

Comment (DHdroid) [reply pyproject.toml:1]: 오잉 이거 제가 add를 잘못했군요

File: scripts/infer_bert.py (hunk 1 above)
Comment (yoonichoi) [reply scripts/infer_bert.py:24]: 넹 해당부분 확인했었는데 저는 이 변경사항 없이도 valid, test inference 잘 돌아갔었거든요. 동현님 돌리셨을때에는 에러가 떴었던건가요?

Comment (DHdroid) [reply scripts/infer_bert.py:24]: 윽.. 아마 그래서 잘못 돌리신 것 같습니다.. train/valid/test가 똑같네요
image

Comment (DHdroid) [reply scripts/infer_bert.py:24]: split을 밖에서 안넣어주면 그냥 기본인 train으로 전부 돌아가게 됩니다

Comment (yoonichoi) [reply scripts/infer_bert.py:24]: a미친 ......... 정신없는 와중에도 작업해서 돌렸건만 훤앛뮤ㅏㄴ우천위추워춰잗ㅊ 다시 다 돌려야것네

Comment (yoonichoi) [reply scripts/infer_bert.py:24]: my badㅠㅠ

Comment (DHdroid) [reply scripts/infer_bert.py:24]: 그래도 TREC은 작아서 뚝딱입니다 ㅋㅋㅋ

File: iclx/inferencer/base_inferencer.py [hunk 3]
```diff
@@ -90,12 +90,12 @@ def _init_model(self, model_name):
         model_config = AutoConfig.from_pretrained(model_name)
         with init_empty_weights():
             empty_model = AutoModelForCausalLM.from_config(model_config)
-        device_map = infer_auto_device_map(empty_model, dtype="float32")
+        device_map = infer_auto_device_map(empty_model, dtype="float16")
```
Comment (nemodleo) [inline iclx/inferencer/base_inferencer.py:93]: float16 은 따로 브랜치 파는건 어떨까요?

Comment (nemodleo) [reply iclx/inferencer/base_inferencer.py:93]: float16에서 결과 이상은 없었나요?

File: iclx/inferencer/base_inferencer.py [hunk 4]
```diff
@@ -90,12 +90,12 @@ def _init_model(self, model_name):
...
         self.model = AutoModelForCausalLM.from_pretrained(model_name,
                                                           device_map=device_map,
                                                           offload_folder="offload",
                                                           offload_state_dict=True,
-                                                          torch_dtype=torch.float32)
+                                                          torch_dtype=torch.float16)
```
Comment (nemodleo) [inline iclx/inferencer/base_inferencer.py:98]: float16.

File: iclx/soft_label_generator/datamodule/ag_news.py
Comment (nemodleo) [inline iclx/soft_label_generator/datamodule/ag_news.py:1]: 다른 데이터모듈 파일은 현재 브랜치 머지 이후 개인들이 다이렉트 추가하는 거를 예상하시나요?

File: iclx/inferencer/base_inferencer.py (hunk 3 above)
Comment (nemodleo) [reply iclx/inferencer/base_inferencer.py:93]: 이상 없음 전달받음.
따로 float16 브랜치 분리 관련 해서 답변만 부탁드립니다

Comment (yoonichoi) [reply iclx/inferencer/base_inferencer.py:93]: 빠른 머지가 관건이다보니 이 한 줄만 바꿔도 되는거면 그냥 요기서 진행하시죠

File: iclx/inferencer/base_inferencer.py (hunk 4 above)
Comment (yoonichoi) [reply iclx/inferencer/base_inferencer.py:98]: 헉 두줄이넹

File: iclx/soft_label_generator/datamodule/ag_news.py
Comment (yoonichoi) [reply iclx/soft_label_generator/datamodule/ag_news.py:1]: 이 PR에서는 그렇게하고 추후 datamodule관련 PR에서 다른 데이터모듈 파일들 일괄수정해서 푸시하는 것 어떨까요? (어차피 HF path 받는걸로 수정해야하니까)

File: iclx/inferencer/base_inferencer.py (hunk 3 above)
Comment (DHdroid) [reply iclx/inferencer/base_inferencer.py:93]: 모두가 같은 코드로 실험 진행하는게 중요할 것 같아서 아예 수정의 여지가 없게 그냥 하드코딩 해놨고, 브랜치는 굳이 나눌 필요 없다고 생각했어요

File: iclx/soft_label_generator/datamodule/ag_news.py
Comment (DHdroid) [reply iclx/soft_label_generator/datamodule/ag_news.py:1]: @nemodleo 네 그렇게 예상하고 있습니다. 어차피 한번 dataset 이름이 fix된 이후부터는 신경 안써도 되는 내용이라 밖에서 argument 받는 것으로 하는게 더 번거로울걸로 생각하고 있어요
